data/ai/micro_ais/micro_ai_helper.lua

```diff
@@ -164,32 +214,42 @@ function micro_ai_helper.micro_ai_setup(cfg, CA_parms, required_keys, optional_k
     local CA_cfg = {}
 
     -- Required keys
-    for _,v in pairs(required_keys) do
-        if v:match('%[[a-zA-Z0-9_]+%]')  then
-            v = v:sub(2,-2)
-            if not wml.get_child(cfg, v) then
-                wml.error("[micro_ai] tag (" .. cfg.ai_type .. ") is missing required parameter: [" .. v .. "]")
+    for k_name,k_type in pairs(required_keys) do
+        if (k_type == 'tag') then
+            -- Check that this is not a scalar parameter
+            if cfg[k_name] then
+                wml.error("[micro_ai] tag (" .. cfg.ai_type .. ") parameter '[" .. k_name .. "]' must be a WML tag")
             end
-            for child in wml.child_range(cfg, v) do
-                table.insert(CA_cfg, T[v](child))
+            if (not wml.get_child(cfg, k_name)) then
+                wml.error("[micro_ai] tag (" .. cfg.ai_type .. ") is missing required parameter: [" .. k_name .. "]")
+            end
+            for child in wml.child_range(cfg, k_name) do
+                table.insert(CA_cfg, T[k_name](child))
             end
         else
-            if not cfg[v] then
-                wml.error("[micro_ai] tag (" .. cfg.ai_type .. ") is missing required parameter: " .. v .."=")
+            check_key_type(k_name, cfg[k_name], k_type, cfg.ai_type)
+            if not cfg[k_name] then
+                wml.error("[micro_ai] tag (" .. cfg.ai_type .. ") is missing required parameter: " .. k_name .."=")
             end
-            CA_cfg[v] = cfg[v]
+            CA_cfg[k_name] = cfg[k_name]
         end
     end
 
     -- Optional keys
-    for _,v in pairs(optional_keys) do
-        if v:match('%[[a-zA-Z0-9_]+%]')  then
-            v = v:sub(2,-2)
-            for child in wml.child_range(cfg, v) do
-                table.insert(CA_cfg, T[v](child))
+    for k_name,k_type in pairs(optional_keys) do
```

Comment:
Okay, sooo… I think this system (both loops) possibly counts as an API, which means this change should go through a deprecation cycle? What this means is essentially that you'd:

1. Add an `if type(k_name) = 'number'` case which uses the original code (where `k_type` is actually the key and `k_name` is ignored).
2. Add a one-time deprecation message in this case. Since this isn't a place where `deprecate_api` is possible, you'd probably do something like (pseudo-code) `if not showed_message then showed_message = true show_message() end`. The `showed_message` variable would be a file-level local variable.